### PR TITLE
change plan to DAG mode

### PR DESCRIPTION
**Pull request by hanfei19910905**

### CODE REVIEW

**Comment by coocood** [comment]
Why `Src` is set as child instead of parent?
And when `planA.AddChilren(planB)`, shouldn't we add `planA` as a parent to planB?

File: optimizer/plan/cost.go [hunk 1]
```diff
@@ -29,49 +29,46 @@ const (
...
 
 // Enter implements Visitor Enter interface.
-func (c *costEstimator) Enter(p Plan) (Plan, bool) {
-	return p, false
-}
-
-// Leave implements Visitor Leave interface.
-func (c *costEstimator) Leave(p Plan) (Plan, bool) {
+func estimate(p Plan) {
+	var child Plan
```
Comment (shenli) [inline optimizer/plan/cost.go]: child will always be the last child of p?

File: optimizer/plan/planbuilder.go [hunk 2]
```diff
@@ -593,8 +593,9 @@ func pushLimit(p Plan, limit *ast.Limit) {
 	case *TableScan:
 		limitCount := int64(limit.Offset + limit.Count)
 		x.LimitCount = &limitCount
-	case WithSrcPlan:
-		pushLimit(x.Src(), limit)
+	default:
+		child := x.GetChildByIndex(0)
```
Comment (shenli) [inline optimizer/plan/planbuilder.go]: check if child == nil?

File: optimizer/plan/cost.go (hunk 1 above)
Comment (hanfei19910905) [inline optimizer/plan/cost.go]: well, here is a hack. cause all the operator needed to be processed only have a single child. cost.go is an incomplete code, will be continued to optimize.

File: optimizer/plan/planbuilder.go (hunk 2 above)
Comment (hanfei19910905) [inline optimizer/plan/planbuilder.go]: will fix it

File: optimizer/plan/plan.go [hunk 3]
```diff
@@ -38,27 +35,18 @@ type Plan interface {
...
 	SetLimit(limit float64)
-}
 
-// WithSrcPlan is a Plan has a source Plan.
-type WithSrcPlan interface {
-	Plan
-	Src() Plan
-	SetSrc(src Plan)
-}
+	AddParent(parent Plan)
```
Comment (shenli) [inline optimizer/plan/plan.go]: Public interface should have comment.

File: optimizer/plan/plans.go [hunk 4]
```diff
@@ -552,51 +314,21 @@ type Delete struct {
...
-	return v.Leave(p)
-}
-
 // SetLimit implements Plan SetLimit interface.
 func (p *Filter) SetLimit(limit float64) {
 	p.limit = limit
-	// We assume 50% of the src row is filtered out.
-	p.src.SetLimit(limit * 2)
+	// We assume 50% of the GetChildByIndex(0) row is filtered out.
+	p.GetChildByIndex(0).SetLimit(limit * 2)
```
Comment (shenli) [inline optimizer/plan/plans.go]: For "select 1 limit 1;", p.GetChildByIndex(0) could be nil?

File: optimizer/plan/refiner.go [hunk 5]
```diff
@@ -24,37 +24,34 @@ import (
...
-	return in, false
-}
-
-func (r *refiner) Leave(in Plan) (Plan, bool) {
+func refine(in Plan) error {
+	var err error
+	for _, c := range in.GetChildren() {
+		e := refine(c)
+		if e != nil {
+			err = e
```
Comment (shenli) [inline optimizer/plan/refiner.go]: return errors.Trace(e)? Or some errors may be override.

File: optimizer/plan/refiner.go [hunk 6]
```diff
@@ -75,22 +72,21 @@ func (r *refiner) buildIndexRange(p *IndexScan) {
...
+		return nil
 	}
 	rb := rangeBuilder{}
 	rangePoints := fullRange
 	for _, cond := range p.AccessConditions {
 		rangePoints = rb.intersection(rangePoints, rb.build(cond))
 	}
 	p.Ranges = rb.buildTableRanges(rangePoints)
-	r.err = rb.err
+	return rb.err
```
Comment (shenli) [inline optimizer/plan/refiner.go]: return errors.Trace(rb.err)

**Comment by hanfei19910905** [comment]
1. because by customary rule,  table scan is the bottom of the tree while data sink is the root of the tree.
2. that's true, but I haven't use any parent ye :)

File: optimizer/plan/plans.go (hunk 4 above)
Comment (hanfei19910905) [inline optimizer/plan/plans.go]: "select 1 limit 1" won't use a filter

File: optimizer/plan/plan.go (hunk 3 above)
Comment (hanfei19910905) [inline optimizer/plan/plan.go]: ok, I will add it

File: optimizer/plan/refiner.go (hunk 5 above)
Comment (hanfei19910905) [inline optimizer/plan/refiner.go]: I will fix it

**Comment by shenli** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Pull request merged**